### PR TITLE
[PageTransitionsBuilder] Fix 'ZoomPageTransition' built more than once

### PR DESCRIPTION
**Pull request by AlexV525**
## Description

This will solve the issue about the `_ZoomPageTransition` cause the widget built three times, and imports the `DualTransitionBuilder` from the animations package.

## Related Issues

Related to #58345 .

## Tests

* `tests/material/page_transition_theme_test.dart`  *_ZoomPageTransition only cause child widget built once*

## Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Did any tests fail when you ran them? Please read [Handling breaking changes].

- [x] No, no existing tests failed, so this is *not* a breaking change.
- [ ] Yes, this is a breaking change.

### CODE REVIEW

**Comment by AlexVincent525** [comment]
The forward transition looks exactly the same as before, but there's still some tricky in the reverse one. Please help me with it to figure how we can implement this more elegantly.

/cc @shihaohong

**Comment by AlexVincent525** [comment]
Oh there're some comments copied from `FadeThroughTransition`, I'll reword them in next commit. 😃

**Comment by shihaohong** [comment]
I'll try to take a look later today or tomorrow. However, if I don't get the chance to, I might disappear for about a week due to personal reasons. In the event that I don't get to it this week, I'll ask someone else on the team to take a look.

Thank you for putting this together and doing the investigation!

**Comment by AlexVincent525** [comment]
When importing the `DualTransitionBuilder`, the `TransitionBuilder` has already defined in `src/widgets/framwork.dart`. Which is different from `Widget Function(BuildContext context, Animation<double> animation, Widget child)`. So I changed its name to `AnimatedTransitionBuilder`.
https://github.com/flutter/flutter/blob/94b7ff241e6e5445b7f30215a777eb4971311797/packages/flutter/lib/src/widgets/framework.dart#L4485

**Comment by AlexVincent525** [comment]
> but there's still some tricky problems in the reverse one and I cannot figure it out. 

For now the problems has been solved and this animation is absolutely the same as before! 🎉 I'll add built times test case ASAP.

**Comment by AlexVincent525** [comment]
Test has been added, this PR is ready for review.

File: packages/flutter/lib/src/widgets/dual_transition_builder.dart
Comment (shihaohong) [inline packages/flutter/lib/src/widgets/dual_transition_builder.dart:19]: We should consider what this means for the copy of `DualTransitionBuilder` in the animations package once this makes it into stable. Would we be able to safely remove `DualTransitionBuilder` from the animations package and use this version instead? 

Also, I don't think this'll be a problem, but I'll ask just in case since I'm not sure what will actually happen -- Will there be any conflicts/problems with having both the copy here and the copy in the animations package exist at the same time?

cc/ @goderbauer

File: packages/flutter/lib/src/widgets/dual_transition_builder.dart
Comment (shihaohong) [inline packages/flutter/lib/src/widgets/dual_transition_builder.dart:5]: We shouldn't import the widgets package in this file since this file is also in that same package. Instead, we should import every file that's needed independently [like this](https://github.com/flutter/flutter/blob/master/packages/flutter/lib/src/widgets/heroes.dart#L7).

File: packages/flutter/lib/src/widgets/dual_transition_builder.dart
Comment (shihaohong) [inline packages/flutter/lib/src/widgets/dual_transition_builder.dart:32]: If we do bring the builder in here, we should bring in the tests from the animations package as well:
https://github.com/flutter/packages/blob/master/packages/animations/test/dual_transition_builder_test.dart

File: packages/flutter/lib/src/material/page_transitions_theme.dart
Comment (shihaohong) [inline packages/flutter/lib/src/material/page_transitions_theme.dart:291]: Did you mean when the animation was _not_ completed?

Comment (shihaohong) [reply packages/flutter/lib/src/material/page_transitions_theme.dart:291]: It might also be valuable to describe how `isCompleted` is used for the `ScaleTransition` as well, since that isn't clearly explained.

File: packages/flutter/lib/src/material/page_transitions_theme.dart
Comment (shihaohong) [inline packages/flutter/lib/src/material/page_transitions_theme.dart:215]: Thanks for adding these API docs!

**Comment by shihaohong** [comment]
The analyzer is [complaining about trailing whitespaces](https://github.com/flutter/flutter/pull/58686/checks?check_run_id=742263871), so those'll need to be addressed

**Comment by AlexVincent525** [comment]
> The analyzer is [complaining about trailing whitespaces](https://github.com/flutter/flutter/pull/58686/checks?check_run_id=742263871), so those'll need to be addressed

This test addressed some import issue that I cannot recognize. 😕

File: packages/flutter/lib/src/widgets/dual_transition_builder.dart
Comment (AlexVincent525) [reply packages/flutter/lib/src/widgets/dual_transition_builder.dart:32]: I'll wait until we have conclusion from Goderbauer about the copy.

**Comment by shihaohong** [comment]
Oops, I misread the error message. It's basically complaining about [this line](https://github.com/flutter/flutter/pull/58686/files#diff-5cf1b6d4f71499b508a1db1f0dd5c890R5) and that it is a recursive dependency

File: packages/flutter/lib/src/material/page_transitions_theme.dart
Comment (goderbauer) [reply packages/flutter/lib/src/material/page_transitions_theme.dart:244]: This also applies to the other widgets defined in this PR.

**Comment by AlexVincent525** [comment]
Working on the review, please wait for a few minutes.

File: packages/flutter/lib/src/material/page_transitions_theme.dart
Comment (goderbauer) [inline packages/flutter/lib/src/material/page_transitions_theme.dart:297]: Just define the statics in the class in which they are used for cleaner encapsulation.

File: packages/flutter/lib/src/material/page_transitions_theme.dart
Comment (goderbauer) [inline packages/flutter/lib/src/material/page_transitions_theme.dart:292]: I don't fully understand this. Why does the opacity jump to 0.0 when the animation is completed?

File: packages/flutter/lib/src/material/page_transitions_theme.dart
Comment (goderbauer) [inline packages/flutter/lib/src/material/page_transitions_theme.dart:304]: Why this? It appears that both tweens produce the same value when the animation is completed?

File: packages/flutter/lib/src/material/page_transitions_theme.dart
Comment (goderbauer) [inline packages/flutter/lib/src/material/page_transitions_theme.dart:293]: The container will have to be wrapped in an AnimatedBuilder since you access the animation's value directly, the container needs to rebuild every time the value changes.

File: packages/flutter/lib/src/material/page_transitions_theme.dart
Comment (goderbauer) [inline packages/flutter/lib/src/material/page_transitions_theme.dart:327]: Same here as above: why do we jump suddenly when the animation is dismissed?

File: packages/flutter/lib/src/widgets/dual_transition_builder.dart
Comment (goderbauer) [reply packages/flutter/lib/src/widgets/dual_transition_builder.dart:32]: +1

File: packages/flutter/lib/src/widgets/dual_transition_builder.dart
Comment (goderbauer) [reply packages/flutter/lib/src/widgets/dual_transition_builder.dart:19]: We should delete it from the animations package and use the one from the framework here. Deleting itself should be fine since the builder was never part of the public interface of the animations package. However while the DualTransitionBuilder exists in two places, I suspect there may be a conflict within the animations package as there are now two competing DualTransitionBuilder definitions. If that is the case, we should hide the framework implementation in the package until this change has made it to stable and then delete the implementation in the package (while increasing the minimal flutter version). 

Can you please check whether the example app in the animation package builds correctly with this change applied to your Flutter version? If it doesn't we will have to hide this implementation in the package before this change can land.

File: packages/flutter/lib/src/material/page_transitions_theme.dart
Comment (goderbauer) [inline packages/flutter/lib/src/material/page_transitions_theme.dart:329]: nit: can you turn this around for easier readability: isDismissed ? kAlwaysCompleteAnimation : foo.

Comment (goderbauer) [reply packages/flutter/lib/src/material/page_transitions_theme.dart:329]: here and elsewhere

File: packages/flutter/lib/src/widgets/dual_transition_builder.dart
Comment (AlexVincent525) [reply packages/flutter/lib/src/widgets/dual_transition_builder.dart:20]: > Can you please check whether the example app in the animation package builds correctly with this change applied to your Flutter version? If it doesn't we will have to hide this implementation in the package before this change can land.

Got it, scheduled for tomorrow since it's already 2.A.M.

File: packages/flutter/lib/src/material/page_transitions_theme.dart
Comment (AlexVincent525) [reply packages/flutter/lib/src/material/page_transitions_theme.dart:336]: > Same here as above: why do we jump suddenly when the animation is dismissed?

I've update the explanation about why the `isCompleted` and `isDismissed` exist, can you check it out to see if it has enough information for the reason?

File: packages/flutter/lib/src/material/page_transitions_theme.dart
Comment (AlexVincent525) [reply packages/flutter/lib/src/material/page_transitions_theme.dart:244]: Done.

Comment (AlexVincent525) [reply packages/flutter/lib/src/material/page_transitions_theme.dart:244]: Done.

File: packages/flutter/lib/src/material/page_transitions_theme.dart
Comment (AlexVincent525) [reply packages/flutter/lib/src/material/page_transitions_theme.dart:302]: Using `AnimatedBuilder` to wrap the `Container` will cause the forward transition failed. (For now idk why it'll happen).

File: packages/flutter/lib/src/material/page_transitions_theme.dart
Comment (AlexVincent525) [reply packages/flutter/lib/src/material/page_transitions_theme.dart:301]: The forward builder stays in the widget tree, so the container needs to be transparent otherwise the child will be all in black.

File: packages/flutter/lib/src/material/page_transitions_theme.dart
Comment (AlexVincent525) [reply packages/flutter/lib/src/material/page_transitions_theme.dart]: Yes but it produce different value when animating. It require different value between the forward end and the reverse start, as the implementation before.

File: packages/flutter/lib/src/material/page_transitions_theme.dart
Comment (AlexVincent525) [reply packages/flutter/lib/src/material/page_transitions_theme.dart:313]: What do mean statics here? The animation needs access in methods...?

**Comment by AlexVincent525** [comment]
AFK for 8 to 10 hours since it's 3.A.M in China, nits were updated. Will be back soon. 😃

**Comment by shihaohong** [comment]
I just checked your branch and it doesn't look quite the same:

This PR:
zoom_transition

Previously: 
previous_appearance

I'm going to see if I can rectify this issue, but one other thing I think I will need to do is add some transition animations tests to ensure that this type of regression is not as easy to make

**Comment by shihaohong** [comment]
Here's my attempt at making the transition value fixes with a PR to @AlexVincent525's feature branch: https://github.com/AlexVincent525/flutter/pull/1/

**Comment by AlexVincent525** [comment]
@shihaohong  I found that the `DualTransitionBuilder` is exported again, is it okay for the animations example?

**Comment by AlexVincent525** [comment]
> I just checked your branch and it doesn't look quite the same:

That's really weird, maybe there's something wrong with my example. I'll double check that.

**Comment by AlexVincent525** [comment]
Yes it's my problem, I simply using the transition in a `PageRouteBuilder`. I'm merging your fix now.

File: packages/flutter/lib/src/material/page_transitions_theme.dart
Comment (AlexVincent525) [inline packages/flutter/lib/src/material/page_transitions_theme.dart:256]: Should this be renamed to `_ZoomEnterTransition` for readability?

File: packages/flutter/lib/src/material/page_transitions_theme.dart
Comment (AlexVincent525) [inline packages/flutter/lib/src/material/page_transitions_theme.dart:317]: Here's the same. `_ExitTransition` -> `_ZoomExitTransition`.

File: packages/flutter/lib/src/material/page_transitions_theme.dart
Comment (AlexVincent525) [inline packages/flutter/lib/src/material/page_transitions_theme.dart:256]: Should this be renamed to _ZoomEnterTransition for readability?

File: packages/flutter/lib/src/material/page_transitions_theme.dart
Comment (AlexVincent525) [inline packages/flutter/lib/src/material/page_transitions_theme.dart:317]: `_ExitTransition` -> `_ZoomExitTransition` ?

File: packages/flutter/lib/widgets.dart
Comment (AlexVincent525) [inline packages/flutter/lib/widgets.dart:38]: If we export it here it'll be public for use.

**Review by AlexV525 — COMMENTED** [review]
@shihaohong I add some questions in my review.

File: packages/flutter/lib/widgets.dart
Comment (AlexVincent525) [reply packages/flutter/lib/widgets.dart:38]: Confirmed, there're conflicts in animations package if we export it.

image

File: packages/flutter/lib/src/widgets/dual_transition_builder.dart
Comment (AlexVincent525) [reply packages/flutter/lib/src/widgets/dual_transition_builder.dart:20]: @goderbauer Confirmed, there're conflicts in animations package if we export it.

image

File: packages/flutter/lib/src/material/page_transitions_theme.dart
Comment (AlexVincent525) [inline packages/flutter/lib/src/material/page_transitions_theme.dart:260]: ❤️ Great implement!

Comment (AlexVincent525) [inline packages/flutter/lib/src/material/page_transitions_theme.dart:260]: Great implement! ❤️

File: packages/flutter/lib/src/material/page_transitions_theme.dart
Comment (shihaohong) [reply packages/flutter/lib/src/material/page_transitions_theme.dart:257]: Sounds good to me

File: packages/flutter/lib/src/material/page_transitions_theme.dart
Comment (shihaohong) [reply packages/flutter/lib/src/material/page_transitions_theme.dart:318]: Sounds good!

File: packages/flutter/test/widgets/dual_transition_builder_test.dart
Comment (shihaohong) [inline packages/flutter/test/widgets/dual_transition_builder_test.dart:300]: nit: newline at the end of a file

**Comment by shihaohong** [comment]
Assigning to @goderbauer for the start of the week, since I'll be out for the next few days

File: packages/flutter/lib/src/material/page_transitions_theme.dart
Comment (goderbauer) [inline packages/flutter/lib/src/material/page_transitions_theme.dart:9]: This is a layer violation. You can't import an implementation file from the widget layer directly into the material layer. The widget layer should export the transition builder so that you can just import `package:flutter/widgets.dart` here.

File: packages/flutter/lib/src/material/page_transitions_theme.dart
Comment (goderbauer) [inline packages/flutter/lib/src/material/page_transitions_theme.dart:263]: assert that `reverse` (and probably also `animation`) is not null.

File: packages/flutter/lib/src/material/page_transitions_theme.dart
Comment (goderbauer) [reply packages/flutter/lib/src/material/page_transitions_theme.dart:306]: This one is still unresolved, as far as I can tell, there is nothing right now ensuring that the Container gets rebuild when the opacity changes.

File: packages/flutter/lib/src/material/page_transitions_theme.dart
Comment (goderbauer) [inline packages/flutter/lib/src/material/page_transitions_theme.dart:292]: A comment should explain why code is written the way it is, it should not paraphrase the code in words like this one is doing. In particular, this comment should explain why the animation status matters (this is still not clear to me).

File: packages/flutter/lib/src/material/page_transitions_theme.dart
Comment (goderbauer) [inline packages/flutter/lib/src/material/page_transitions_theme.dart:325]: Same asserts here as above.

File: packages/flutter/lib/src/material/page_transitions_theme.dart
Comment (goderbauer) [inline packages/flutter/lib/src/material/page_transitions_theme.dart:350]: nit: fix indentation

File: packages/flutter/lib/src/material/page_transitions_theme.dart
Comment (goderbauer) [inline packages/flutter/lib/src/material/page_transitions_theme.dart:351]: Instead of `!reverse` express this positively: `reverse` to simplify things.

**Comment by goderbauer** [comment]
> Confirmed, there're conflicts in animations package if we export it.

We need to hide the framework's implementation in the animations package then until we have a stable framework release with the transition builder so we can upgrade the animations package to use it.

File: packages/flutter/lib/src/material/page_transitions_theme.dart
Comment (AlexVincent525) [reply packages/flutter/lib/src/material/page_transitions_theme.dart:9]: Oh I thought that we need hide from here. Sorry for that.

File: packages/flutter/lib/src/material/page_transitions_theme.dart
Comment (AlexVincent525) [reply packages/flutter/lib/src/material/page_transitions_theme.dart:351]: @goderbauer Is it good to use
```dart
(reverse
  ? _scaleDownTransition
  : _scaleUpTransition
).animate(animation)
```
Or
```dart
reverse
  ? _scaleDownTransition.animate(animation)
  : _scaleUpTransition.animate(animation)
```

File: packages/flutter/lib/src/material/page_transitions_theme.dart
Comment (AlexVincent525) [reply packages/flutter/lib/src/material/page_transitions_theme.dart:292]: It's really hard for a person whose native language is not English 😕. I'm wondering if this looks okay:
```dart
// When the transition is running with the [reverseBuilder] in [DualTransitionBuilder],
// the status of the animation will be always [AnimationStatus.completed].
//
// To ensure the opacity only changed with forward animation, we need to combine
// the [reverse] flag with the animation status.
```

Comment (AlexVincent525) [reply packages/flutter/lib/src/material/page_transitions_theme.dart:293]: @goderbauer

**Comment by AlexVincent525** [comment]
Now this PR was affected by the non-nullable check, what should I do?

**Comment by shihaohong** [comment]
@AlexVincent525 regarding the nnbd stuff, you need to add a comment to your new files [just like this](https://github.com/flutter/flutter/blob/master/packages/flutter/lib/src/material/mergeable_material.dart#L5). Here's a [context PR](https://github.com/flutter/flutter/pull/59186)

**Comment by shihaohong** [comment]
Also, thank you for your patience, I have been busy with moving home recently and finally had some time to settle back into a more regular schedule.

File: packages/flutter/lib/src/material/page_transitions_theme.dart
Comment (shihaohong) [inline packages/flutter/lib/src/material/page_transitions_theme.dart:299]: Thanks for improving the wording here, the original I left in here was really vague. Let me try to clarify this more, you can use what you think makes the most sense from it and add parts of yours that you think is helpful as well: 
```
    // The transition's scrim opacity only increases on the forward transition. In the reverse
    // transition, the opacity should always be 0.0.
    //
    // Therefore, we need to only apply the scrim opacity animation when the transition
    // is running forwards.
```

cc/ @goderbauer I think I tried `if (!reverse && animation.status == AnimationStatus.forward)` but it did not behave as I expected, but I cannot remember the exact details of the weird behavior.

File: packages/flutter/lib/src/widgets/dual_transition_builder.dart
Comment (shihaohong) [inline packages/flutter/lib/src/widgets/dual_transition_builder.dart:4]:
```suggestion

// @dart = 2.8

```

File: packages/flutter/test/widgets/dual_transition_builder_test.dart
Comment (shihaohong) [inline packages/flutter/test/widgets/dual_transition_builder_test.dart:4]:
```suggestion

// @dart = 2.8

```

File: packages/flutter/lib/src/material/page_transitions_theme.dart
Comment (AlexVincent525) [reply packages/flutter/lib/src/material/page_transitions_theme.dart:299]: I'll take your wording, because the reason I've written which related to the `DualTransitionBuilder` seems unnecessary here.

File: packages/flutter/lib/src/material/page_transitions_theme.dart
Comment (Hixie) [inline packages/flutter/lib/src/material/page_transitions_theme.dart:154]: the docs should also include a description of the constructor (or you can change `///` to `//` to indicate it's not camera-ready documentaion)

Comment (AlexVincent525) [reply packages/flutter/lib/src/material/page_transitions_theme.dart:154]:
```dart
/// Creates a [_ZoomPageTransition].
///
/// The [animation] and [secondaryAnimation] argument are required and must
/// not be null.
```

Something like this?

Comment (AlexVincent525) [reply packages/flutter/lib/src/material/page_transitions_theme.dart:154]: Or we can remove these comments since the transition is a hidden one.

File: packages/flutter/lib/src/material/page_transitions_theme.dart
Comment (shihaohong) [reply packages/flutter/lib/src/material/page_transitions_theme.dart]: I looked into it again for the reason it has to be `if (!reverse && animation.status != AnimationStatus.completed)`. This allows an interrupted reversal of the forward transition to smoothly fade the scrim away.

Using `if (!reverse && animation.status == AnimationStatus.forward)` causes the scrim to instantly vanish if the animation was interrupted while it was transitioning forwards. This results in a disjointed experience, so it's likely better to have the scrim fade away rather than having it suddenly disappear.

**Review by shihaohong — APPROVED** [review]
LGTM modulo the constructor API doc comment

File: packages/flutter/lib/src/material/page_transitions_theme.dart
Comment (AlexVincent525) [reply packages/flutter/lib/src/material/page_transitions_theme.dart]: So should I take further action with this comment?

Comment (shihaohong) [reply packages/flutter/lib/src/material/page_transitions_theme.dart]: I was thinking of waiting for what goderbauer thought, but I think it also makes sense to add a few sentences explaining this:
```
    // The transition's scrim opacity only increases on the forward transition. In the reverse
    // transition, the opacity should always be 0.0.
    //
    // Therefore, we need to only apply the scrim opacity animation when the transition
    // is running forwards.
    //
    // The reason that this we check that the animation's status is not `completed` instead
    // of checking that it is `forward` is that this allows the interrupted reversal of the
    // forward transition to smoothly fade the scrim away. This prevents a disjointed 
    // removal of the scrim.
```

Comment (AlexVincent525) [reply packages/flutter/lib/src/material/page_transitions_theme.dart]: Then I'll hold off the commit since there are already 31 commits in this PR. But he has disappeared for a week. 🤣

Comment (shihaohong) [reply packages/flutter/lib/src/material/page_transitions_theme.dart]: The number of commits shouldn’t matter for the PR since we squash and merge (everything shows up as a single commit).

Comment (AlexVincent525) [reply packages/flutter/lib/src/material/page_transitions_theme.dart]: Okay then I'm going to mark this as resolved.

**Comment by AlexVincent525** [comment]
Seems all resolved now, is it good to go?

**Comment by shihaohong** [comment]
You still need to address @Hixie's comment right? Sorry, I think my wording might have been confusing. I was saying that this change you suggested looks okay:
```
/// Creates a [_ZoomPageTransition].
///
/// The [animation] and [secondaryAnimation] argument are required and must
/// not be null.
```

Unless there's something else that needs to be described in the constructor that I might be missing

**Comment by AlexVincent525** [comment]
Oh that is because I dropped one of the commit which is included. 😢

**Comment by AlexVincent525** [comment]
> Sorry, I think my wording might have been confusing.

It's my bad, done it.

**Comment by AlexVincent525** [comment]
Tree is green now  👀

**Comment by shihaohong** [comment]
@AlexVincent525 It seems one of the test suites timed out. I restarted it and also applied the `waiting for tree to green` label, which will merge the PR once all checks are green.

Thank you so much for this contribution, it was a lot of work and I'm glad we got through it :)

**Pull request merged**
